PWM updates for STM32

rsalveti opened the pull request:
* Add STM32 PWM 3 (TIM3)
* Add PWM3 channel definitions on PC 6, 7, 8 and 9
* Fix check for APB prescale

rsalveti committed:
pwm: stm32: Add PWM 3 (TIM3)

Add support for TIM3 as it is widely available and pins available via
headers on several devices.

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>

rsalveti committed:
pinmux: stm32: stm32f4: Add PWM3 channels on PC 6, 7, 8 and 9

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>

vaussard commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -140,20 +119,16 @@ static int pwm_stm32_get_cycles_per_sec(struct device *dev, u32_t pwm,
-			(clock_control_subsys_t *)&cfg->pclken);
-
-	*cycles = (u64_t)(tim_clk / (data->pwm_prescaler + 1));
+	*cycles = (u64_t)(bus_clk / (data->pwm_prescaler + 1));
```
No, this is wrong. But on the other side, there is also a bug in __get_tim_clk(), so...
If you look in the "Timer clock" section under RCC in the datasheet (6.2.15 in my case for a STM32L4x6), the timer clock is
1. equal to the bus clock if APB prescaler is 1
2. twice the bus clock otherwise

Thus you miss the second case. Try with an oscilloscope and change PPRE1 or PPRE2 depending on your timer.

rsalveti replied:
Indeed, was looking at the bus clock but the timer clock can indeed be twice the bus clock.

rsalveti replied:
Found out what was causing my confusion, there is nothing saying that the timer clock needs to take the APB prescale into account in the STM32F401RE manual (only "The timers connected to APB2 are clocked from TIMxCLK up to 84 MHz, while the timers connected to APB1 are clocked from TIMxCLK up to 42 MHz").

Now checking at the other F/L manuals, they all state what you just described above. Also took a look at stm32cube for F4xx and the same sentence is covered in the pwm sample readme.txt.

vaussard commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -26,27 +26,6 @@
-		apb_psc = CONFIG_CLOCK_STM32_APB2_PRESCALER;
-	}
-
-	if (apb_psc == RCC_HCLK_DIV1) {
```
There is a bug here. The RCC_HCLK_DIV1 macro will resolve to 0x00 (value of the PPRE1/2 register). But on the other side, CONFIG_CLOCK_STM32_APB[12]_PRESCALER will use a value of 1 (or more) if we want a prescaler of one (or more). Thus the condition is always false. RCC_HCLK_DIV1 must be replaced by the literal "0x01" instead or a similar macro.

I have a patch for that in my tree, but had no time to upstream the changes. So feel free.

rsalveti replied:
Will have a look and update the patch, thanks for the quick review.

rsalveti committed:
pwm: stm32: Fix check for APB prescale

RCC_HCLK_DIV1 translates to 0x0 while apb_psc uses the value defined
by CONFIG_CLOCK_STM32_APB1/2_PRESCALER (range from 1 to 16).

Manually check if the defined prescaler is 1 or not and use that to
calculate the correct timer clock.

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>

rsalveti commented:
@vaussard couldn't find the relevant patch in your public tree, but I replaced the patch with one changing the code as you suggested.

vaussard commented:
Indeed, it is not in my public tree  But the patch is trivial :)

rsalveti commented:
@galak should I propose "pwm: stm32: Fix check for APB prescale" in a separated pr? That would probably be a good fix for 1.9.

galak commented:
@rsalveti the 'Fix check for APB prescale' do you feel that should go into 1.9.1?

rsalveti commented:
> @rsalveti the 'Fix check for APB prescale' do you feel that should go into 1.9.1?

Yes, since the current logic is known to be broken. Will you just cherry-pick?
